### PR TITLE
Fix typing in napari.layers.base

### PR DESCRIPTION
**Pull request by dstansby**
Builds on top of napari/napari#5892.

We might need extra work on typing later on, see the following comment: https://github.com/napari/napari/pull/6028/files#r1284169897, 

But his will be easier once all the layers have typing enabled.

### CODE REVIEW

File: napari/layers/base/base.py
Comment (Carreau) [inline napari/layers/base/base.py:1736]: Is that True ? Because this new one implies that `(None, ndarray)` is a value that can be returned, while before, it was not the case.

Comment (Carreau) [reply napari/layers/base/base.py:1736]: Maybe for later in another PR, the docs should update the returns section to say it can be None.

Comment (Carreau) [reply napari/layers/base/base.py:1736]: From what I can tell `intersect_line_with_axis_aligned_bounding_box_3d` cannot return None, so either both are None, or only one is None.

File: napari/layers/base/base.py
Comment (Carreau) [inline napari/layers/base/base.py:1673]: Same concern (None, None) pair here than below.

**Comment by dstansby** [comment]
Thanks for bringing that up, I should have explained in a comment. It seems `mypy` cannot 'link' two variables, so for a function:
```
def f() -> Union[Tuple[np.ndarray, np.ndarray], Tuple[None, None]]:
    return a, b
```
Both `a` and `b` will have type `np.ndarray | None`. This typing is fine for the orginal function (`f()`) above, but if `a, b` is returned from a second function, the second function has to have typing `Tuple[np.ndarray | None, np.ndarray | None`. 

I think the only round this is altering the typing as I have in this PR, I tried various forms of using `cast()` but none of them seem to make `mypy` happy.

**Comment by Carreau** [comment]
Ah ok, that make sens. I hope mypy will be able to recognize this pattern later.

File: napari/layers/labels/_labels_key_bindings.py
Comment (Carreau) [inline napari/layers/labels/_labels_key_bindings.py]: This seem to be the line that cause errors. 
This is typically why I think relative import are better, as you can make sure that you don't import from cousins' submodules, and have a tree and not a DAG. But it seen napari prefers absolute imports.

I think we can use `if TYPE_CHECKING:` and use a str in the `cast`

Comment (Carreau) [reply napari/layers/labels/_labels_key_bindings.py]: I pushed the commit to fix the circular import.

**Comment by Carreau** [comment]
Hang on `napari/_qt/widgets/_tests/test_qt_dims_2.py`. Restarted.

When test hang and someone restart them, please note in a comment which test hang so we can keep track of hanging and investigate recurrent hanging.

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py]: This shouldn't be just `SceneOverlay` but simply `Overlay`, it's just that currently no layer overlays are `CanvasOverlay`.

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py]: Why is this needed instead of the simple `self._modecalss.PAN_ZOOM`?

File: napari/layers/base/base.py
Comment (brisvag) [reply napari/layers/base/base.py:1736]: ~Yeah this looks wrong... Maybe we need to fix the actual implementation somewhere?~ Sorry, saw the answer afterwards.

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py:1311]: Is this really right? Surely these inputs are needed?

File: napari/layers/labels/_labels_key_bindings.py
Comment (brisvag) [inline napari/layers/labels/_labels_key_bindings.py]: can you explain this? I'd rather avoid obscure type casting stuf just for the sake of making mypy happy :/

**Review by brisvag — DISMISSED** [review]

File: napari/layers/base/base.py
Comment (Carreau) [reply napari/layers/base/base.py]: We need to make the super class a proper generics on `Mode`.
This is one of the only way to make mypy not complain that both:
1) self._modeclass  does not have attribute XXX
2) later on the XXXX is for sure not None.

Other way is to put an # type: ignore.

It's both a  quirk of mypy and the fact that generics are still recent and quite complex to implement so will be a PR of their own. 

see https://github.com/napari/napari/issues/6005 for followup on generics.

File: napari/layers/base/base.py
Comment (Carreau) [reply napari/layers/base/base.py:1311]: I believe these are passed directly from `get_ray_intersections`, that returns optional. 
Maybe we can put assert those are not None on the I belive the only place where it is called ? (line 1253 a bit above this one).

File: napari/layers/base/base.py
Comment (Carreau) [reply napari/layers/base/base.py:1673]: Marking as resolved as this is a mypy quirk.

File: napari/layers/base/base.py
Comment (Carreau) [reply napari/layers/base/base.py:1311]: Also I think this has no implementation, should it become (in a future PR) and abstract method. 

Also those 5 tests fail if we assert those 2 parameters are not None: 


```
napari/layers/points/_tests/test_points.py                   test_value_3d[position2-view_direction2-dims_displayed2-False-scale2-None]
napari/layers/points/_tests/test_points.py                   test_value_3d[position7-view_direction7-dims_displayed7-True-scale7-None]
napari/layers/points/_tests/test_points_mouse_bindings.py    test_selecting_points_with_drag_3d
napari/layers/points/_tests/test_points_mouse_bindings.py    test_selecting_no_points_with_drag_3d
napari/layers/shapes/_tests/test_shapes.py                   test_value_3d[position2-view_direction2-dims_displayed2-False-scale2-None]
```

So they do get `None` when rays do not intersect I guess.

Comment (brisvag) [reply napari/layers/base/base.py:1311]: Ah, I see, that makes sense.

Comment (Carreau) [reply napari/layers/base/base.py:1311]: Marking as resolved then.

File: napari/layers/labels/_labels_key_bindings.py
Comment (Carreau) [reply napari/layers/labels/_labels_key_bindings.py]: `layer._overlays` is defined as an EventedDict containing `Overlay`s in values on the base class.

add_polygon_to_labels is only defined on the subclass LabelsPolygonOverlay

We could use `typing.TypedDict` but those are not typed, we could have a `TypedEventedDict`.

So this is really just to please mypy.

We could open an issue for EventedTypedDict, or just have a custom object as we don't have that many keys for this dict. 

```
'polygon'
'axes'
'bounding_box'
'brush_circle'
'dot'
'scale_bar'
'text'
'transform_box'
```

File: napari/layers/base/base.py
Comment (Carreau) [reply napari/layers/base/base.py]: I see a +1 from you so i'm going to mark as resolved.

File: napari/layers/labels/_labels_key_bindings.py
Comment (brisvag) [reply napari/layers/labels/_labels_key_bindings.py]: I *really* don't like having a hard to read, more verbose, and more complicated code just to make mypy happy and not because it's actually useful to have types there :/ can't we just ignore this?

Comment (Carreau) [reply napari/layers/labels/_labels_key_bindings.py]: I guess we could use  `type: ignore`, the issue with ignore is that it may hide other issues on the given line. I don't have much preferences.

Comment (dstansby) [reply napari/layers/labels/_labels_key_bindings.py]: 👍 I added a type ignore and a comment with TODO so hopefully someone can look at this again in the future

File: napari/layers/base/base.py
Comment (dstansby) [reply napari/layers/base/base.py]: 👍 done

**Review by Carreau — APPROVED** [review]

**Comment by dstansby** [comment]
Needs some more updates after https://github.com/napari/napari/pull/6082 was merged, so converted to draft.

**Comment by dstansby** [comment]
Should be good again now 😄

File: napari/layers/base/base.py
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
        PAN_ZOOM = getattr(self._modeclass, "PAN_ZOOM")
        TRANSFORM = getattr(self._modeclass, "TRANSFORM")
```
Why not this? Based on python [documentation](https://docs.python.org/3/library/functions.html#getattr) this approach will return a more readable error.

**Review by Czaki — APPROVED** [review]
Looks good, one small question. When it was explained I think it will be shortly label as ready to merge.

File: napari/layers/base/base.py
Comment (dstansby) [reply napari/layers/base/base.py]: I'm not sure, and your way is definitely better 😄

File: napari/layers/base/base.py
Comment (Carreau) [inline napari/layers/base/base.py]:
```suggestion
        PAN_ZOOM = self._modeclass.PAN_ZOOM # type: ignore[attr-defined]
        TRANSFORM = self._modeclass.TRANSFORM # type: ignore[attr-defined]
```

I think that was the reason for the getattr/assert None. Let's try with type ignore. 
TO properly fix that we'll need Generics over Modeclass, and we are not there yet.

**Review by Carreau — APPROVED** [review]

**Review by Czaki — APPROVED** [review]

**Comment by Carreau** [comment]
Resolved conflicts, it was two imports added in the same line, so fairly stritforward.

**Comment by Carreau** [comment]
Removing "ready to merge" typing failed after merging with master and there was a could changes like removing default arguments to some private functions. Re-requesting review.

File: napari/layers/base/base.py
Comment (Czaki) [inline napari/layers/base/base.py]: maybe use name `order_array` and `order`? 
it may be more readable than numbers.

**Review by Czaki — APPROVED** [review]
One small wording, but maybe someone else has opinion abou this.

File: napari/layers/base/base.py
Comment (Carreau) [reply napari/layers/base/base.py]: Done.

**Review by brisvag — APPROVED** [review]

**Comment by dstansby** [comment]
🎉 thanks everyone for helping get this over the line!

**Comment by Carreau** [comment]
Thanks for stating the movement.

**Comment by melonora** [comment]
Thanks @dstansby for all the typing work:) Please if anything is hanging feel free to send me a message on zulip.

**Pull request merged**
